parsec/core/types/manifest.py

```diff
@@ -450,22 +457,144 @@ def match_remote(self, remote_manifest: RemoteFileManifest) -> bool:
+            name: entry_id
+            for name, entry_id in other.children.items()
+            if entry_id in other.confined_entries
+        }
+        return self.evolve_children_and_mark_updated(previously_confined_entries, pattern_filter)
+
+    def _filter_remote_entries(
+        self: LocalFolderishManifestTypeVar, pattern_filter: Pattern
+    ) -> LocalFolderishManifestTypeVar:
+        filtered_entries = frozenset(
+            {entry_id for name, entry_id in self.children.items() if pattern_filter.match(name)}
+        )
+        if not filtered_entries:
+            return self
+        children = {
+            name: entry_id
+            for name, entry_id in self.children.items()
+            if entry_id not in filtered_entries
+        }
+        return self.evolve(children=children, filtered_entries=filtered_entries)
+
+    def _restore_filtered_entries(
+        self: LocalFolderishManifestTypeVar
+    ) -> LocalFolderishManifestTypeVar:
+        if not self.filtered_entries:
+            return self
+        children = dict(self.children)
+        for name, entry_id in self.base.children.items():
+            if entry_id in self.filtered_entries:
+                children[name] = entry_id
+        return self.evolve(filtered_entries=frozenset(), children=children)
+
+    # Apply filter
+
+    def apply_filter(
+        self: LocalFolderishManifestTypeVar, pattern_filter: Pattern
+    ) -> LocalFolderishManifestTypeVar:
+        # Filter confined entries
+        result = self._filter_confined_entries()
+        # Restore filtered entries
+        result = result._restore_filtered_entries()
+        # Filter remote entries
+        result = result._filter_remote_entries(pattern_filter)
+        # Restore confined entries
+        return result._restore_confined_entries(self, pattern_filter)
+
+
 @attr.s(slots=True, frozen=True, auto_attribs=True, kw_only=True, eq=False)
-class LocalFolderManifest(BaseLocalManifest):
+class LocalFolderManifest(BaseLocalManifest, LocalFolderishManifestMixin):
     class SCHEMA_CLS(BaseSchema):
         type = fields.EnumCheckedConstant(LocalManifestType.LOCAL_FOLDER_MANIFEST, required=True)
         base = fields.Nested(RemoteFolderManifest.SCHEMA_CLS, required=True)
         need_sync = fields.Boolean(required=True)
         updated = fields.DateTime(required=True)
         children = fields.FrozenMap(EntryNameField(), EntryIDField(required=True), required=True)
+        confined_entries = fields.FrozenSet(EntryIDField(required=True))
+        filtered_entries = fields.FrozenSet(EntryIDField(required=True))
 
         @post_load
         def make_obj(self, data):
             data.pop("type")
+            data.setdefault("confined_entries", frozenset())
+            data.setdefault("filtered_entries", frozenset())
             return LocalFolderManifest(**data)
 
     base: RemoteFolderManifest
     children: FrozenDict[EntryName, EntryID]
+    confined_entries: FrozenSet[EntryID]
```

Comment:
Should put a comment here about the difference between confined and filtered entries